go/teams/service_helper.go

```diff
@@ -436,65 +436,102 @@ type AddMembersRes struct {
-				normalizedUsername = username
+				return err
 			}
-			res[i] = AddMembersRes{
-				Invite:   invite,
-				Username: normalizedUsername,
+
+			tx := CreateAddMemberTx(team)
+			type sweepEntry struct {
+				Assertion string
+				UV        keybase1.UserVersion
 			}
-			if !uv.IsNil() {
-				sweep = append(sweep, sweepEntry{
-					Assertion: user.AssertionOrEmail,
-					UV:        uv,
+			var sweep []sweepEntry
+			for _, user := range users {
+				upak, single, doInvite, err := tx.ResolveUPKV2FromAssertionOrEmail(mctx, user.AssertionOrEmail)
+				if err != nil {
+					return NewAddMembersError(user.AssertionOrEmail, err)
+				}
+
+				if _, ok := restrictedUsers[upak.Uid]; ok {
+					// skip users with contact setting restrictions
+					user := keybase1.User{Uid: upak.Uid, Username: libkb.NewNormalizedUsername(upak.Username).String()}
+					notAdded = append(notAdded, user)
+					continue
+				}
+
+				username, uv, invite, err := tx.AddOrInviteMemberByUPKV2(ctx, upak, single, doInvite, user.AssertionOrEmail, user.Role, user.BotSettings)
+				if err != nil {
+					if _, ok := err.(AttemptedInviteSocialOwnerError); ok {
+						return err
+					}
+					return NewAddMembersError(user.AssertionOrEmail, err)
+				}
+				var normalizedUsername libkb.NormalizedUsername
+				if !username.IsNil() {
+					normalizedUsername = username
+				}
+
+				added = append(added, AddMembersRes{
+					Invite:   invite,
+					Username: normalizedUsername,
 				})
+				if !uv.IsNil() {
+					sweep = append(sweep, sweepEntry{
+						Assertion: user.AssertionOrEmail,
+						UV:        uv,
+					})
+				}
 			}
+			// Try to mark completed any invites for the users' social assertions.
+			// This can be a time-intensive process since it involves checking proofs.
+			// It is limited to a few seconds and failure is non-fatal.
+			timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 2*time.Second)
+			for _, x := range sweep {
+				if err := tx.CompleteSocialInvitesFor(timeoutCtx, x.UV, x.Assertion); err != nil {
+					g.Log.CWarningf(ctx, "Failed in CompleteSocialInvitesFor(%v, %v) -> %v", x.UV, x.Assertion, err)
+				}
+			}
+			timeoutCancel()
+
+			return tx.Post(libkb.NewMetaContext(ctx, g))
 		}
+	}
 
-		// Try to mark completed any invites for the users' social assertions.
-		// This can be a time-intensive process since it involves checking proofs.
-		// It is limited to a few seconds and failure is non-fatal.
-		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 2*time.Second)
-		for _, x := range sweep {
-			if err := tx.CompleteSocialInvitesFor(timeoutCtx, x.UV, x.Assertion); err != nil {
-				g.Log.CWarningf(ctx, "Failed in CompleteSocialInvitesFor(%v, %v) -> %v", x.UV, x.Assertion, err)
-			}
+	// try to add
+	restrictedUsers := make(map[keybase1.UID]bool)
+	err = RetryIfPossible(ctx, g, retryableFunc(restrictedUsers))
+	if isTeamContactSettingsBlock(err) {
+		// retry add
+		fields := err.(libkb.AppStatusError).Fields
+		uids := fields["uids"]
```

Comment:
@mmou did you see my comment that this should be already done in libkb/rpc_exim? If it's not working let's figure out why

Reply:
did you see my reply here https://github.com/keybase/client/pull/21528#discussion_r356962516